### PR TITLE
New sim data

### PR DESCRIPTION
**Pull request by fscarlier**
I added a new class and function to be able to keep the full 6D position data (and s, E) from the trackone files. Decided to create separate class and function so not to touch the TransverseData class which is marked for measurement data, and avoid problems downstream.

### CODE REVIEW

**Review by fsoubelet — COMMENTED** [review]
Very nice. Could you add a test as done for the current `trackone` reading in `test_ptc_trackone.py`?

You should be able to use the same test files as they have all the columns.

File: tests/test_lhc_and_general.py [hunk 1]
```diff
@@ -56,12 +56,13 @@ def test_tbt_write_read_ascii(_sdds_file, _test_file):
...
 
-def compare_tbt(origin: TbtData, new: TbtData, no_binary: bool, max_deviation=ASCII_PRECISION) -> None:
+def compare_tbt(origin: TbtData, new: TbtData, no_binary: bool, max_deviation = ASCII_PRECISION, full_sim_data: bool = False) -> None:
+    FIELDS = PLANES if not full_sim_data else SIMDATA_FIELDS  # check in ("X", "Y") if just trackone otherwise all fields
     assert new.nturns == origin.nturns
     assert new.nbunches == origin.nbunches
     assert new.bunch_ids == origin.bunch_ids
     for index in range(origin.nbunches):
-        for plane in PLANES:
+        for plane in FIELDS:
```
Comment (JoschD) [inline tests/test_lhc_and_general.py:65]: `parameter` ? `field`?

File: turn_by_turn/structures.py [hunk 2]
```diff
@@ -30,14 +30,38 @@ def __getitem__(self, item):  # to access X and Y like one would with a dictiona
         return getattr(self, item)
 
 
+@dataclass
+class SimulationData:
```
Comment (JoschD) [inline turn_by_turn/structures.py:34]: 8D-Data?

File: turn_by_turn/constants.py [hunk 3]
```diff
@@ -7,6 +7,7 @@
 from typing import Dict, Tuple
 
 PLANES: Tuple[str, str] = ("X", "Y")
+SIMDATA_FIELDS: Tuple[str, str, str, str, str, str, str, str] = ("X", "PX", "Y", "PY", "T", "PT", "S", "E")  # fields in SimulationData
```
Comment (JoschD) [inline turn_by_turn/constants.py:10]: thanks. I hate it ^^

File: turn_by_turn/trackone.py [hunk 4]
```diff
@@ -14,25 +14,32 @@
...
 
 from turn_by_turn.structures import TbtData
-from turn_by_turn.utils import numpy_to_tbt
+from turn_by_turn.utils import numpy_to_sim_tbt, numpy_to_tbt
 
 LOGGER = logging.getLogger()
 
 
-def read_tbt(file_path: Union[str, Path]) -> TbtData:
+def read_tbt(file_path: Union[str, Path], full_sim_data: bool = False) -> TbtData:
```
Comment (JoschD) [inline turn_by_turn/trackone.py:22]: `full data` ?

File: turn_by_turn/utils.py [hunk 5]
```diff
@@ -176,3 +176,36 @@ def numpy_to_tbt(names: np.ndarray, matrix: np.ndarray) -> TbtData:
...
+    Returns:
+        A ``TbtData`` object loaded with the matrices in the provided numpy arrays.
+    """
+    # get list of TbTFile from 4D matrix ...
+    _, _, nbunches, nturns = matrix.shape
+    matrices = []
+    indices = []
+    for index in range(nbunches):
+        matrices.append(
+            SimulationData(
```
Comment (JoschD) [inline turn_by_turn/utils.py:199]: **{field: pd.DataFrame(index=names, data=matrix[idx, :, index, :]) for idx, field in enumerate(SIMDATA_FIELDS)}

File: turn_by_turn/structures.py [hunk 6]
```diff
@@ -30,14 +30,38 @@ def __getitem__(self, item):  # to access X and Y like one would with a dictiona
...
+    X: pd.DataFrame  # horizontal data
+    PX: pd.DataFrame  # horizontal momentum data
+    Y: pd.DataFrame  # vertical data
+    PY: pd.DataFrame  # vertical momentum data
+    T: pd.DataFrame  # longitudinal data
+    PT: pd.DataFrame  # longitudinal momentum data
+    S: pd.DataFrame  # longitudinal position data
+    E: pd.DataFrame  # energy data
+
+    def fieldnames(self):
```
Comment (JoschD) [inline turn_by_turn/structures.py:48]: Why not use this, instead of SIMDATA_FIELDS?

**Review by JoschD — COMMENTED** [review]
Some suggestions

**Comment by JoschD** [comment]
https://github.com/pylhc/turn_by_turn/blob/ef52f3a96f274a2ed34b4b30f1cf7df265c065e9/turn_by_turn/utils.py#L21

Should also be adapted to go through the fields of the underlying tbt data instead of X,Y

File: turn_by_turn/utils.py (hunk 5 above)
Comment (JoschD) [reply turn_by_turn/utils.py:199]: also, this function could have the datatype as input:
```
DataType = Union[TransverseData, SimulationData]

def numpy_to_tbt(names: np.ndarray, matrix: np.ndarray, data_type: DataType = TransverseData) -> TbtData:
```

File: tests/test_lhc_and_general.py (hunk 1 above)
Comment (fsoubelet) [reply tests/test_lhc_and_general.py:65]: field is good!

File: turn_by_turn/constants.py (hunk 3 above)
Comment (fsoubelet) [reply turn_by_turn/constants.py:10]: Me too

File: turn_by_turn/structures.py (hunk 2 above)
Comment (fsoubelet) [reply turn_by_turn/structures.py:34]: Maybe `TrackingData`?

File: turn_by_turn/structures.py (hunk 6 above)
Comment (fsoubelet) [reply turn_by_turn/structures.py:48]: Sounds good

File: turn_by_turn/trackone.py (hunk 4 above)
Comment (fsoubelet) [reply turn_by_turn/trackone.py:22]: I like `is_tracking_data`

**Review by JoschD — APPROVED** [review]

**Pull request merged**